lib/Dialect/Arc/Transforms/SplitLoops.cpp

```diff
@@ -0,0 +1,358 @@
+    valueColoring.insert(
+        {operand.get(),
+         APInt::getOneBitSet(numResults, operand.getOperandNumber())});
+
+  for (auto &op : llvm::reverse(defOp.getBodyBlock().without_terminator())) {
+    auto coloring = APInt::getZero(numResults);
+    for (auto result : op.getResults())
+      if (auto it = valueColoring.find(result); it != valueColoring.end())
+        coloring |= it->second;
+    opColoring.insert({&op, coloring});
+    op.walk([&](Operation *op) {
+      for (auto &operand : op->getOpOperands())
+        valueColoring.try_emplace(operand.get(), numResults, 0).first->second |=
+            coloring;
+    });
+  }
+
+  // Determine the splits for this arc.
+  Splitter splitter(&getContext(), defOp.getLoc());
+  splitter.run(defOp.getBodyBlock(), opColoring);
+
+  // Materialize the split arc definitions.
+  ImplicitLocOpBuilder builder(defOp.getLoc(), defOp);
+  SmallVector<DefineOp> splitArcs;
+  splitArcs.reserve(splitter.splits.size());
+  for (auto &split : splitter.splits) {
+    auto splitName = defOp.getSymName();
+    if (splitter.splits.size() > 1)
+      splitName = arcNamespace.newName(defOp.getSymName() + "_split_" +
+                                       Twine(split->index));
+    auto splitArc = builder.create<DefineOp>(
+        splitName, builder.getFunctionType(
+                       split->block->getArgumentTypes(),
+                       split->block->getTerminator()->getOperandTypes()));
+    splitArc.getBody().push_back(split->block.release());
+    splitArcs.push_back(splitArc);
+  }
+
+  // Replace all uses with the new splits and remove the old definition.
+  for (auto arcUse : arcUses)
+    replaceArcUse(arcUse, splitArcs, splitter.splits, splitter.outputs);
+  defOp.erase();
+}
+
+/// Replace a use of the original arc with new uses for the splits.
+void SplitLoopsPass::replaceArcUse(StateOp arcUse, ArrayRef<DefineOp> splitDefs,
+                                   ArrayRef<Split *> splits,
+                                   ArrayRef<ImportedValue> outputs) {
+  ImplicitLocOpBuilder builder(arcUse.getLoc(), arcUse);
+  SmallVector<StateOp> newUses;
+
+  // Resolve an `ImportedValue` to either an operand of the original arc or the
+  // result of another split.
+  auto getMappedValue = [&](ImportedValue value) {
+    if (value.isInput)
+      return arcUse.getInputs()[value.index];
+    return newUses[value.split].getResult(value.index);
+  };
+
+  // Collect the operands for each split and create a new use for each. These
+  // are either operands of the original arc, or values from other splits
+  // exported as results.
+  for (auto [splitDef, split] : llvm::zip(splitDefs, splits)) {
+    SmallVector<Value> operands;
+    for (auto importedValue : split->importedValues)
+      operands.push_back(getMappedValue(importedValue));
+    auto newUse =
+        builder.create<StateOp>(splitDef, Value{}, Value{}, 0, operands);
+    newUses.push_back(newUse);
```

Comment:
Nit: I guess preallocating the right amount of elements and indexing here according to the split index would be more robust?